### DISCUSSION TITLE
Bus error when running GDL on macOS Sonoma 14.2.1

### DISCUSSION THREAD

**Discussion by spacemanjosh** [gnudatalanguage/gdl#1728] [Q&A] [answered]
First time building GDL, so I may be missing something.  Any help is appreciated.
I've been able to build GDL with the steps outlined here: [https://github.com/gnudatalanguage/gdl/wiki/GDL-on-OSX](https://github.com/gnudatalanguage/gdl/wiki/GDL-on-OSX)
I'm building in $HOME/gdl
But when I run even the simplest IDL commands I get a hard crash.  The only information printed to the console indicates a bus error:

```
GDL> x = 42
GDL> print, x
zsh: bus error  ./install/bin/gdl
```

The build steps I took are:

```
cd $HOME
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
brew install cmake
brew install gcc
git clone https://github.com/gnudatalanguage/gdl.git
cd gdl
./scripts/build_gdl.sh prep
./scripts/build_gdl.sh configure
./scripts/build_gdl.sh build
./scripts/build_gdl.sh install
```

I've also tried building the app and dmg, which seems to work fine, and I'm able to launch GDL via the app icon as well.  But I still get the crash either way.

I'm likely missing something

**Comment by alaingdl** [comment]
:disappointed: 
Could you give more informations ? OS version (`sw_vers`), type of processor ...

On OSX I made frequent & extensive tests on x86 (intel) and also M1/M2 without any troubles
(except the specific performance issues  #1498 especially 
https://github.com/gnudatalanguage/gdl/issues/1498#issuecomment-1846296486

**Reply by spacemanjosh** [reply]
Sure, so I'm running on an M2 Max chip, macOS 14.2.1.  It's entirely possible I'm just missing something in the build process, or something else in my environment.

I wonder if it's the version of gcc.  I notice that although I'm putting the homebrew bin directory first on my path, a `which gcc` seems to still point to the system copy.  How sensitive to the version of gcc is the build process?

**Comment by alaingdl** [comment]
I tested (two weeks ago) on M2 & macOS 14 (not sure on the details) and it was fine (including plot through network !)

1/ did you make : ./scripts/build_gdl.sh check (or test)
2/ do you have conda/miniconde/anaconda somewhere in the path ?

could you have a look to : otools -L /path/to/exe/gdl (maybe something is missing ?)  
(sorry if a type, cannot check now)

**Reply by spacemanjosh** [reply]
I had removed conda from my path prior to the build.  I just double checked, and we're all good there.
As for the `./scripts/build_gdl.sh check`, all of the tests return a Bus error.  
`otool -L ./install/bin/gdl` doesn't seem to indicate anything missing.

As a test, I was able to run gdl on a Fedora instance, which works great there.  I'm not sure what's wrong with my macOS build though.

**Comment by alaingdl** [comment]
@spacemanjosh Ha, I am not on 14.x, but only 13.y ... Maybe the main point ... No way to test that by myself

```
% sw_vers 
ProductName:		macOS
ProductVersion:		13.4.1
ProductVersionExtra:	(c)
BuildVersion:		22F770820d
```

(remotely on a laptop of one student here ... thanks !)

Whouaou, large update of the Brew when making ./scripts/build_gdl.sh prep, including LLVM ...
now compiling ! ... success ... all tests OK ...

Difficult to help more :disappointed: 

What I would do in your case : after carefully check that nothing related to Conda remains, compile again
(removing before the build path (rm -rf build) in order to clean the cache)


```
100% tests passed, 0 tests failed out of 205

Total Test time (real) = 113.82 sec

The following tests did not run:
	135 - test_mpi.pro (Skipped)
	154 - test_python.pro (Skipped)

```

(and plots are working remotely)
```
GDL> plot, findgen(10)
GDL> !gdl
{
    "RELEASE": "1.0.4 git",
    "BUILD_DATE": "Jan 19 2024",
    "EPOCH": 1705618800,
    "GDL_USE_DSFMT": 1,
    "GDL_USE_WX": 0,
    "GDL_POSIX": 1
}
GDL> !version
{
    "ARCH": "arm64",
    "OS": "darwin",
    "OS_FAMILY": "unix",
    "OS_NAME": "Mac OS X",
    "RELEASE": "8.2",
    "BUILD_DATE": "Jan 19 2024",
    "MEMORY_BITS": 64,
    "FILE_OFFSET_BITS": 64
}

alain@macm2 gdl-1.0.4git240119 % otool -L ./install/bin/gdl
./install/bin/gdl:
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.100.3)
	/usr/lib/libncurses.5.4.dylib (compatibility version 5.4.0, current version 5.4.0)
	/usr/lib/libform.5.4.dylib (compatibility version 5.4.0, current version 5.4.0)
	/opt/homebrew/opt/readline/lib/libreadline.8.dylib (compatibility version 8.2.0, current version 8.2.0)
	/opt/homebrew/opt/readline/lib/libhistory.8.dylib (compatibility version 8.2.0, current version 8.2.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/opt/homebrew/opt/gsl/lib/libgsl.27.dylib (compatibility version 28.0.0, current version 28.0.0)
	/opt/homebrew/opt/gsl/lib/libgslcblas.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libplplot.17.dylib (compatibility version 17.0.0, current version 17.0.0)
	@rpath/libplplotcxx.15.dylib (compatibility version 15.0.0, current version 15.0.0)
	/opt/homebrew/opt/libpng/lib/libpng16.16.dylib (compatibility version 57.0.0, current version 57.0.0)
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	/System/Library/Frameworks/Carbon.framework/Versions/A/Carbon (compatibility version 2.0.0, current version 169.0.0)
	/System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa (compatibility version 1.0.0, current version 23.0.0)
	/System/Library/Frameworks/QuartzCore.framework/Versions/A/QuartzCore (compatibility version 1.2.0, current version 1.11.0)
	/System/Library/Frameworks/AudioToolbox.framework/Versions/A/AudioToolbox (compatibility version 1.0.0, current version 1000.0.0)
	/System/Library/Frameworks/OpenGL.framework/Versions/A/OpenGL (compatibility version 1.0.0, current version 1.0.0)
	/opt/homebrew/opt/wxwidgets/lib/libwx_baseu-3.2.dylib (compatibility version 3.0.0, current version 3.2.0)
	/opt/homebrew/opt/wxwidgets/lib/libwx_osx_cocoau_core-3.2.dylib (compatibility version 3.0.0, current version 3.2.0)
	/opt/homebrew/opt/libx11/lib/libX11.6.dylib (compatibility version 11.0.0, current version 11.0.0)
	/opt/X11/lib/libXext.6.dylib (compatibility version 11.0.0, current version 11.0.0)
	/opt/homebrew/opt/graphicsmagick/lib/libGraphicsMagick.3.dylib (compatibility version 28.0.0, current version 28.4.0)
	/opt/homebrew/opt/graphicsmagick/lib/libGraphicsMagick++.12.dylib (compatibility version 20.0.0, current version 20.0.0)
	/opt/homebrew/opt/libtiff/lib/libtiff.6.dylib (compatibility version 7.0.0, current version 7.2.0)
	/opt/homebrew/opt/libgeotiff/lib/libgeotiff.5.dylib (compatibility version 8.0.0, current version 8.0.0)
	/opt/homebrew/opt/netcdf/lib/libnetcdf.19.dylib (compatibility version 19.0.0, current version 19.0.0)
	/opt/homebrew/opt/hdf5/lib/libhdf5.310.dylib (compatibility version 314.0.0, current version 314.0.0)
	/opt/homebrew/opt/libaec/lib/libsz.2.dylib (compatibility version 2.0.0, current version 2.0.1)
	/opt/homebrew/opt/open-mpi/lib/libmpi.40.dylib (compatibility version 81.0.0, current version 81.1.0)
	/opt/homebrew/opt/fftw/lib/libfftw3.3.dylib (compatibility version 10.0.0, current version 10.10.0)
	/opt/homebrew/opt/fftw/lib/libfftw3f.3.dylib (compatibility version 10.0.0, current version 10.10.0)
	/opt/homebrew/opt/proj/lib/libproj.25.dylib (compatibility version 25.0.0, current version 25.9.3)
	/opt/homebrew/opt/udunits/lib/libudunits2.0.dylib (compatibility version 2.0.0, current version 2.0.0)
	/opt/homebrew/opt/eccodes/lib/libeccodes.dylib (compatibility version 0.0.0, current version 0.0.0)
	/opt/homebrew/opt/glpk/lib/libglpk.40.dylib (compatibility version 44.0.0, current version 44.1.0)
	/opt/homebrew/opt/shapelib/lib/libshp.4.dylib (compatibility version 4.0.0, current version 1.6.0)
	/usr/lib/libexpat.1.dylib (compatibility version 7.0.0, current version 8.0.0)
	/opt/homebrew/opt/llvm/lib/libomp.dylib (compatibility version 5.0.0, current version 5.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1500.65.0)
	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/ApplicationServices (compatibility version 1.0.0, current version 61.0.0)
	/System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics (compatibility version 64.0.0, current version 1690.5.4)


```

**Reply by spacemanjosh** [reply]
Still no luck here.  Everything else looks ok. I even went so far as to completely remove miniconda from my system, but no luck.  As a test, I tried building it in a container running Fedora, and that seems to work just fine.  That should get me by for now, although it's a little less than ideal.  If anyone else has tested on macOS 14.x I would be curious to see if they have the same problem.  Again, this could all be something I'm doing incorrectly.

**Comment by Barney-9000** [comment, marked as answer]
Could [#1675](https://github.com/gnudatalanguage/gdl/discussions/1675#discussion-5862744) be the solution?  Worked for me (Sonoma 14.2.1) about six weeks ago.

**Reply by spacemanjosh** [reply]
Thanks!  That seems to have worked.  I believe it was the openmp flag you mention.  All I had to do was this from the `gdl/build` directory:
`cmake -DOPENMP=OFF .. `
And now the executable seems to work, without crashing.

I then ran `./scripts/build_gdl.sh check`, which mostly worked, except for a single failure (and two skipped):

```
The following tests did not run:
	135 - test_mpi.pro (Skipped)
	154 - test_python.pro (Skipped)

The following tests FAILED:
	131 - test_matrix_multiply.pro (Failed)
```

The output from that test specifically is:

```
131/205 Test #131: test_matrix_multiply.pro ...........***Failed    0.67 sec
% Compiled module: TEST_MATRIX_MULTIPLY.
% TEST_MATRIX_MULTIPLY: Types 14 and 15 desactivated due to pbs in INVERT() in GDL for these 2 type
% Compiled module: DIAG_MATRIX.
size of matrix :        2
nb errors :        4
input type :        6
effective type :            6
size of matrix :        3
nb errors :        4
input type :        6
effective type :            6
size of matrix :        4
nb errors :        4
input type :        6
effective type :            6
% TEST_MATRIX_MULTIPLY: At least one error encountered, please rerun code with /verbose
```

So it's possible there still may be something not quite right, but this is definitely a step in the right direction.

**Reply by GillesDuvert** [reply]
@spacemanjosh apparently there is a problem with complex (matrix) multiplication (type 6). As there are no other problems signaled, it would seem that only this type, on this architecture, and probably only with the Eigen:: library used behind the scene, will be affected. 
This should be reported as an issue, but complementary tests will  be needed.

**Reply by spacemanjosh** [reply]
That's certainly interesting @GillesDuvert, I'd be happy to file an issue.  But first I'd be curious if others on macOS Sonoma also see that test failing.  @Barney-9000, does test 131 fail for you as well?

**Comment by Barney-9000** [comment]
Actually, it passed -- note the test is numbered 130:

```
        Start 130: test_matrix_multiply.pro
130/202 Test #130: test_matrix_multiply.pro ...........   Passed    0.67 sec
```

However, test_systime.pro failed:
 
```
Total Test time (real) =  92.59 sec

The following tests did not run:
	134 - test_mpi.pro (Skipped)
	153 - test_python.pro (Skipped)

The following tests FAILED:
	182 - test_systime.pro (Failed)
Errors while running CTest
make: *** [test] Error 8
```

Here's the full listing of the error messages:

```
182/202 Test #182: test_systime.pro ...................***Failed    0.27 sec
% Compiled module: TEST_SYSTIME.
LC_ALL=C: Command not found.
% TEST_SYSTIME_LOCALE: Problem with LOCALE in GDL
% Compiled module: BANNER_FOR_TESTSUITE.
% Compiled module: GDL_IDL_FL.
% TEST_SYSTIME_LOCALE:   1 errors encountered during TEST_SYSTIME_LOCALE tests  
% Compiled module: ERRORS_CUMUL.
% TEST_SYSTIME_ZERO:   NO errors encountered during TEST_SYSTIME_ZERO tests  
% TEST_SYSTIME_EPOCH:   NO errors encountered during TEST_SYSTIME_EPOCH tests  
% TEST_SYSTIME_ELAPSED:   NO errors encountered during TEST_SYSTIME_ELAPSED tests  
% TEST_SYSTIME_JULIAN: Not finished
% TEST_SYSTIME_JULIAN: Not finished, please contributed !
% TEST_SYSTIME_JULIAN: Not finished
% TEST_SYSTIME_JULIAN:   NO errors encountered during TEST_SYSTIME_JULIAN tests  
% TEST_SYSTIME: ====================================================
% TEST_SYSTIME: =                                                  =
% TEST_SYSTIME: =  1 errors encountered during TEST_SYSTIME tests  =
% TEST_SYSTIME: =                                                  =
% TEST_SYSTIME: ====================================================
```

My version of gdl was compiled in December on an M2 Ultra Studio under (I think) Sonoma 14.1, however the test carried out just now was under 14.3.

```
GDL> !gdl
{
    "RELEASE": "1.0.4 git",
    "BUILD_DATE": "Dec 19 2023",
    "EPOCH": 1702944000,
    "GDL_USE_DSFMT": 1,
    "GDL_USE_WX": 1,
    "GDL_POSIX": 1
```

**Comment by GillesDuvert** [comment]
I've access to an apple mini with Apple silicon M1, and it reports the same error as @spacemanjosh .
This really gonna be an issue.
